### PR TITLE
Canvas inspector hover

### PR DESCRIPTION
**Pull request by huwshimi**
Highlight service tokens when services in the inspector are hovered or selected.

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- add a few services
- in the inspector hover over the services
- the service token should get the hover state
- view a service in the inspector
- the matching service token should get selected

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3663/
Test PASSed.

**Comment by hatched** [comment]
QA Issues:
- Hovering over subordinates in the added services list doesn't trigger a state change on the subordinate on the canvas (deployed the django bundle) 
- Hovering over the canvas service icons don't put the added-services-list-item into a hover state
- If you hover over python-django, in added services list, click it, then click to get back to the added services list, python-django on the canvas is still in 'hover' state until you hover over the python-django line item again.

**Comment by huwshimi** [comment]
@hatched Those issues should be resolved. I'm not 100% sure about the approach for syncing the token hover back to the inspector. Could you take a look and see what you think?

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3684/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3685/
Test PASSed.

File: jujugui/static/gui/src/app/app.js [hunk 1]
```diff
@@ -798,22 +798,36 @@ YUI.add('juju-gui', function(Y) {
...
+      // Set up a handler for syncing the service token hover. This needs to be
+      // attached only when the component is visible.
+      if (this.hoverService) {
+        this.hoverService.detach();
+      }
+      this.hoverService = topo.on('hoverService', function(service) {
+        this._renderAddedServices(service.id);
+      }, this);
+      // Deselect the active service token.
+      ServiceModule.deselectNodes();
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: why is this necessary?

File: jujugui/static/gui/src/app/app.js [hunk 2]
```diff
@@ -798,22 +798,36 @@ YUI.add('juju-gui', function(Y) {
...
       var utils = views.utils;
       var db = this.db;
+      var topo = this.views.environment.instance.topo;
+      var ServiceModule = topo.modules.ServiceModule;
+      // Set up a handler for syncing the service token hover. This needs to be
+      // attached only when the component is visible.
+      if (this.hoverService) {
+        this.hoverService.detach();
+      }
+      this.hoverService = topo.on('hoverService', function(service) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: Is there a reason why this is inside here? Because this method is called frequently you have to detach and attach the event all the time (as you already know) I feel like this would make more sense in the initializer?

File: jujugui/static/gui/src/app/app.js [hunk 3]
```diff
@@ -934,6 +950,12 @@ YUI.add('juju-gui', function(Y) {
         document.getElementById('charmbrowser-container'));
     },
 
+    _emptySectionA: function() {
+      if (this.hoverService) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: see my comment about the attach point above

**Comment by hatched** [comment]
This functionality is so awesome
QA :ok:

File: jujugui/static/gui/src/app/app.js (hunk 2 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/app.js]: This has to be here because if it is in the init it will try and render the added services component if the user hovers a service when they have the service details open.

File: jujugui/static/gui/src/app/app.js (hunk 1 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/app.js]: When a user navigates to the service details the service token will highlight, then when they close the service details `ServiceModule.deselectNodes();` makes the service token deactivate again.

Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: Sounds good could you add a code comment to this effect

File: jujugui/static/gui/src/app/app.js (hunk 2 above)
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: Sounds good could you add a code comment to this effect

**Comment by hatched** [comment]
Thanks for the reply :+1:

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3689/
Test PASSed.

**Comment by kadams54** [comment]
Looks great to me! :+1: and QA is :ok:.

**Comment by hatched** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/juju-gui-merge/1255

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge
